pkg/snmp/traps/formatter.go

```diff
@@ -0,0 +1,249 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package traps
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/gosnmp/gosnmp"
+)
+
+// Formatter is an interface to extract and format raw SNMP Traps
+type Formatter interface {
+	FormatPacket(packet *SnmpPacket) ([]byte, error)
+	GetTags(packet *SnmpPacket) []string
+}
+
+// JSONFormatter is a Formatter implementation that transforms Traps into JSON
+type JSONFormatter struct {
+	oidResolver OIDResolver
+	namespace   string
+}
+
+type trapVariable struct {
+	OID     string      `json:"oid"`
+	VarType string      `json:"type"`
+	Value   interface{} `json:"value"`
+}
+
+const (
+	sysUpTimeInstanceOID = "1.3.6.1.2.1.1.3.0"
+	snmpTrapOID          = "1.3.6.1.6.3.1.1.4.1.0"
+)
+
+// NewJSONFormatter creates a new JSONFormatter instance with an optional OIDResolver variable.
+func NewJSONFormatter(oidResolver OIDResolver) JSONFormatter {
+	namespace := GetNamespace()
+	if oidResolver == nil {
+		return JSONFormatter{NoOpOIDResolver{}, namespace}
+	}
+	return JSONFormatter{oidResolver, namespace}
+}
+
+// FormatPacket converts a raw SNMP trap packet to a FormattedSnmpPacket containing the JSON data and the tags to attach
+func (f JSONFormatter) FormatPacket(packet *SnmpPacket) ([]byte, error) {
+	var formattedData map[string]interface{}
```

Comment:
I think it can help to defined a struct for the trap payload instead of using generic `map[string]interface{}`. WDYT?

Example:

```
type TrapsPayload struct {
	TrapOID        string `json:"trap_oid"`
	EnterpriseOID        string `json:"enterprise_oid"`
        [...]
	Variables      trapVariable
}
```

Reply:
I wanted to do that but it's unfortunately not possible as the TrapPayload struct would need an undetermined amount of fields, with names that are not determined in advance because of the resolving.
Unless we store both the TrapPayload as you defined and a generic `map[string]interface{}` that we latter combine into a single json blob, but that doesn't really solve the problem and it introduces more complexity